java/integTest/src/test/java/glide/standalone/PubSubTests.java

```diff
@@ -0,0 +1,827 @@
+                (msg, ctx) -> ((ConcurrentLinkedDeque<Pair<Integer, Message>>) ctx).push(Pair.of(1, msg));
+
+        var listener =
+                useCallback
+                        ? createClientWithSubscriptions(
+                                standalone, subscriptions, Optional.of(callback), Optional.of(mq))
+                        : createClientWithSubscriptions(standalone, subscriptions);
+        var sender = createClient(standalone);
+        clients.addAll(List.of(listener, sender));
+
+        assertEquals(1L, sender.publish(channel, message).get());
+        Thread.sleep(404); // deliver the message
+
+        verifyReceivedMessages(
+                Set.of(Pair.of(1, new Message(message, channel))), listener, useCallback);
+    }
+
+    @SneakyThrows
+    @SuppressWarnings("unchecked")
+    @ParameterizedTest(name = "standalone = {0}, use callback = {1}")
+    @MethodSource("getTwoBoolPermutations")
+    public void exact_happy_path_many_channels(boolean standalone, boolean useCallback) {
+        int numChannels = 256;
+        int messagesPerChannel = 256;
+        var messages = new ArrayList<Message>(numChannels * messagesPerChannel);
+        ChannelMode mode = standalone ? PubSubChannelMode.EXACT : PubSubClusterChannelMode.EXACT;
+        Map<? extends ChannelMode, Set<String>> subscriptions = Map.of(mode, new HashSet<>());
+
+        for (var i = 0; i < numChannels; i++) {
+            var channel = UUID.randomUUID().toString();
+            subscriptions.get(mode).add(channel);
+            for (var j = 0; j < messagesPerChannel; j++) {
+                var message = UUID.randomUUID().toString();
+                messages.add(new Message(message, channel));
+            }
+        }
+
+        MessageCallback callback =
+                (msg, ctx) -> ((ConcurrentLinkedDeque<Pair<Integer, Message>>) ctx).push(Pair.of(1, msg));
+
+        var listener =
+                useCallback
+                        ? createClientWithSubscriptions(
+                                standalone, subscriptions, Optional.of(callback), Optional.of(mq))
+                        : createClientWithSubscriptions(standalone, subscriptions);
+        var sender = createClient(standalone);
+        clients.addAll(List.of(listener, sender));
+
+        for (var message : messages) {
+            assertEquals(1L, sender.publish(message.getChannel(), message.getMessage()).get());
+        }
+
+        Thread.sleep(404); // deliver the messages
+
+        verifyReceivedMessages(
+                messages.stream().map(m -> Pair.of(1, m)).collect(Collectors.toSet()),
+                listener,
+                useCallback);
+    }
+
+    @SneakyThrows
+    @SuppressWarnings("unchecked")
+    @ParameterizedTest(name = "use callback = {0}")
+    @ValueSource(booleans = {true, false})
+    public void sharded_pubsub(boolean useCallback) {
+        assumeTrue(REDIS_VERSION.isGreaterThanOrEqualTo("7.0.0"), "This feature added in redis 7");
```

Comment:
Redis -> Valkey?

Reply:
This goes to the test report only. Valkey <7 doesn't exist. Still need to update that?